<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Add marker genes to reference dendrogram for tree mapping — addDendrogramMarkers • scrattch.taxonomy</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Add marker genes to reference dendrogram for tree mapping — addDendrogramMarkers"><meta name="description" content="Add marker genes to reference dendrogram for tree mapping"><meta property="og:description" content="Add marker genes to reference dendrogram for tree mapping"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">scrattch.taxonomy</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.2</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/go_to_examples.html">Go to examples</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Add marker genes to reference dendrogram for tree mapping</h1>

      <div class="d-none name"><code>addDendrogramMarkers.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Add marker genes to reference dendrogram for tree mapping</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">addDendrogramMarkers</span><span class="op">(</span></span>
<span>  <span class="va">AIT.anndata</span>,</span>
<span>  mode.name <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  mode <span class="op">=</span> <span class="va">AIT.anndata</span><span class="op">$</span><span class="va">uns</span><span class="op">$</span><span class="va">mode</span>,</span>
<span>  celltypeColumn <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  subsample <span class="op">=</span> <span class="fl">100</span>,</span>
<span>  num.markers <span class="op">=</span> <span class="fl">20</span>,</span>
<span>  de.param <span class="op">=</span> <span class="fu">scrattch.hicat</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scrattch.hicat/man/de_param.html" class="external-link">de_param</a></span><span class="op">(</span>low.th <span class="op">=</span> <span class="fl">0.1</span>, padj.th <span class="op">=</span> <span class="fl">1</span>, lfc.th <span class="op">=</span> <span class="fl">0.01</span>, q1.th <span class="op">=</span></span>
<span>    <span class="fl">0.1</span>, q2.th <span class="op">=</span> <span class="cn">NULL</span>, q.diff.th <span class="op">=</span> <span class="fl">0.1</span>, de.score.th <span class="op">=</span> <span class="fl">1</span>, min.cells <span class="op">=</span> <span class="fl">1</span>, min.genes <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>,</span>
<span>  calculate.de.genes <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  save.shiny.output <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  mc.cores <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  bs.num <span class="op">=</span> <span class="fl">100</span>,</span>
<span>  p <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>  low.th <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>  overwriteMarkers <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  taxonomyDir <span class="op">=</span> <span class="fu"><a href="file.path.html">file.path</a></span><span class="op">(</span><span class="va">AIT.anndata</span><span class="op">$</span><span class="va">uns</span><span class="op">$</span><span class="va">taxonomyDir</span><span class="op">)</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-ait-anndata">AIT.anndata<a class="anchor" aria-label="anchor" href="#arg-ait-anndata"></a></dt>
<dd><p>A reference taxonomy anndata object.</p></dd>


<dt id="arg-mode-name">mode.name<a class="anchor" aria-label="anchor" href="#arg-mode-name"></a></dt>
<dd><p>Taxonomy mode to determine which version of filtering to use.</p></dd>


<dt id="arg-celltypecolumn">celltypeColumn<a class="anchor" aria-label="anchor" href="#arg-celltypecolumn"></a></dt>
<dd><p>Column name correspond to the cell type names in the dendrogram. At least two cells per cell type in the dendrogram must be included. Default is to take the highest resolution level of the hierarchy</p></dd>


<dt id="arg-subsample">subsample<a class="anchor" aria-label="anchor" href="#arg-subsample"></a></dt>
<dd><p>The number of cells to retain per cluster (default = 100)</p></dd>


<dt id="arg-num-markers">num.markers<a class="anchor" aria-label="anchor" href="#arg-num-markers"></a></dt>
<dd><p>The maximum number of markers to calculate per pairwise differential calculation per direction (default = 20)</p></dd>


<dt id="arg-de-param">de.param<a class="anchor" aria-label="anchor" href="#arg-de-param"></a></dt>
<dd><p>Differential expression (DE) parameters for genes and clusters used to define marker genes.  By default the values are set to the 10x nuclei defaults from scrattch.hicat, except with min.cells=2 (see notes below).</p></dd>


<dt id="arg-calculate-de-genes">calculate.de.genes<a class="anchor" aria-label="anchor" href="#arg-calculate-de-genes"></a></dt>
<dd><p>Default=TRUE. If set to false, the function will search for "de_genes" in a file called "de_genes_<a href="https://rdrr.io/r/base/mode.html" class="external-link">mode</a>.RData" in the folder taxonomyDir, and will use those if available.</p></dd>


<dt id="arg-save-shiny-output">save.shiny.output<a class="anchor" aria-label="anchor" href="#arg-save-shiny-output"></a></dt>
<dd><p>Should standard output files be generated and saved to the AIT file (default=TRUE).  These are not required for tree mapping, but are required for building a patch-seq shiny instance. See notes.</p></dd>


<dt id="arg-mc-cores">mc.cores<a class="anchor" aria-label="anchor" href="#arg-mc-cores"></a></dt>
<dd><p>Number of cores to use for running this function to speed things up.  Default = 1.  Values&gt;1 are only supported in an UNIX environment and require <code>foreach</code> and <code>doParallel</code> R libraries.</p></dd>


<dt id="arg-bs-num">bs.num<a class="anchor" aria-label="anchor" href="#arg-bs-num"></a></dt>
<dd><p>Number of bootstrap runs for creating the dendrogram (default of 100)</p></dd>


<dt id="arg-p">p<a class="anchor" aria-label="anchor" href="#arg-p"></a></dt>
<dd><p>proportion of marker genes to include in each iteration of the mapping algorithm.</p></dd>


<dt id="arg-low-th">low.th<a class="anchor" aria-label="anchor" href="#arg-low-th"></a></dt>
<dd><p>the minimum difference in Pearson correlation required to decide on which branch</p></dd>


<dt id="arg-overwritemarkers">overwriteMarkers<a class="anchor" aria-label="anchor" href="#arg-overwritemarkers"></a></dt>
<dd><p>If markers already are calculated a tree, should they be overwritten (default = TRUE)</p></dd>


<dt id="arg-taxonomydir">taxonomyDir<a class="anchor" aria-label="anchor" href="#arg-taxonomydir"></a></dt>
<dd><p>The location to create the directory with taxonomy information. Both the taxonomy and the de.genes_mode.name.RData get saved here.</p>
<p>By default VERY loose parameters are set for de_param in an effort to get extra marker genes for each node.  The defaults previously proposed for 10x nuclei are the following <code>de_param(low.th = 1, padj.th = 0.01, lfc.th = 1, q1.th = 0.3, q2.th = NULL, q.diff.th = 0.7, de.score.th = 100, min.cells = 2, min.genes = 5)</code>. See the function <code>de_param</code> in the scrattch.hicat for more details.</p>
<p>If save.shiny.output=TRUE, the following two values will get saved to the uns:
<code>memb.ref</code>   - matrix indicating how much confusion there is the mapping between each cell all of the nodes in the tree (including all cell types) when comparing clustering and mapping results with various subsamplings of the data
<code>map.df.ref</code> - Result of tree mapping for each cell in the reference against the clustering tree, including various statistics and marker gene evidence.  This is the same output that comes from tree mapping.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>An updated dendrogram variable that is the same as <code>dend</code> except with marker genes added to each node.</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Inkar Kapen.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

