<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>This function builds the minimum files required for Shiny Precomputed clusters must be provided. In the anndata object these will be stored using the term "cluster". If hierarchy[-1] is anything other than cluster, then any existing "cluster" column will be overwritten by hierarchy[-1]. Values can be provided without colors and ids (e.g., "cluster") or with them (e.g., "cluster_label" + "cluster_color" + "cluster_id"). In this case cluster_colors is ignored and colors are taken directly from the metadata. Cluster_id's will be overwritten to match dendrogram order. (NOTE: Some functionality with metadata colors is still under development.) — buildTaxonomy • scrattch.taxonomy</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="This function builds the minimum files required for Shiny Precomputed clusters must be provided. In the anndata object these will be stored using the term " cluster if hierarchy is anything other than then any existing column will be overwritten by values can provided without colors and ids or with them in this case cluster_colors ignored are taken directly from the metadata. cluster_id to match dendrogram order. some functionality metadata still under development. buildtaxonomy><meta name="description" content="This function builds the minimum files required for Shiny
Precomputed clusters must be provided.  In the anndata object these will be stored using the term &quot;cluster&quot;.  If hierarchy[-1] is anything other than cluster, then any existing &quot;cluster&quot; column will be overwritten by hierarchy[-1].  Values can be provided without colors and ids (e.g., &quot;cluster&quot;) or with them (e.g., &quot;cluster_label&quot; + &quot;cluster_color&quot; + &quot;cluster_id&quot;).  In this case cluster_colors is ignored and colors are taken directly from the metadata.  Cluster_id's will be overwritten to match dendrogram order.  (NOTE: Some functionality with metadata colors is still under development.)"><meta property="og:description" content="This function builds the minimum files required for Shiny
Precomputed clusters must be provided.  In the anndata object these will be stored using the term &quot;cluster&quot;.  If hierarchy[-1] is anything other than cluster, then any existing &quot;cluster&quot; column will be overwritten by hierarchy[-1].  Values can be provided without colors and ids (e.g., &quot;cluster&quot;) or with them (e.g., &quot;cluster_label&quot; + &quot;cluster_color&quot; + &quot;cluster_id&quot;).  In this case cluster_colors is ignored and colors are taken directly from the metadata.  Cluster_id's will be overwritten to match dendrogram order.  (NOTE: Some functionality with metadata colors is still under development.)"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">scrattch.taxonomy</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.2</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/go_to_examples.html">Go to examples</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>This function builds the minimum files required for Shiny Precomputed clusters must be provided. In the anndata object these will be stored using the term "cluster". If hierarchy[-1] is anything other than cluster, then any existing "cluster" column will be overwritten by hierarchy[-1]. Values can be provided without colors and ids (e.g., "cluster") or with them (e.g., "cluster_label" + "cluster_color" + "cluster_id"). In this case cluster_colors is ignored and colors are taken directly from the metadata. Cluster_id's will be overwritten to match dendrogram order. (NOTE: Some functionality with metadata colors is still under development.)</h1>

      <div class="d-none name"><code>buildTaxonomy.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function builds the minimum files required for Shiny
Precomputed clusters must be provided.  In the anndata object these will be stored using the term "cluster".  If hierarchy[-1] is anything other than cluster, then any existing "cluster" column will be overwritten by hierarchy[-1].  Values can be provided without colors and ids (e.g., "cluster") or with them (e.g., "cluster_label" + "cluster_color" + "cluster_id").  In this case cluster_colors is ignored and colors are taken directly from the metadata.  Cluster_id's will be overwritten to match dendrogram order.  (NOTE: Some functionality with metadata colors is still under development.)</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">buildTaxonomy</span><span class="op">(</span></span>
<span>  title <span class="op">=</span> <span class="st">"AIT"</span>,</span>
<span>  <span class="va">meta.data</span>,</span>
<span>  <span class="va">hierarchy</span>,</span>
<span>  counts <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  normalized.expr <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  highly_variable_genes <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  marker_genes <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  ensembl_id <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  gene.meta.data <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  cluster_stats <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  embeddings <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  number.of.pcs <span class="op">=</span> <span class="fl">30</span>,</span>
<span>  dend <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  taxonomyDir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/getwd.html" class="external-link">getwd</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  cluster_colors <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  default_embedding <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  uns.variables <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  subsample <span class="op">=</span> <span class="fl">2000</span>,</span>
<span>  reorder.dendrogram <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  add.dendrogram.markers <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  addMapMyCells <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  check.taxonomy <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  print.messages <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-title">title<a class="anchor" aria-label="anchor" href="#arg-title"></a></dt>
<dd><p>The file name to assign for the Taxonomy h5ad (default="AIT"; recommended to create your own title!).</p></dd>


<dt id="arg-meta-data">meta.data<a class="anchor" aria-label="anchor" href="#arg-meta-data"></a></dt>
<dd><p>Meta.data corresponding to count matrix. Rownames must be equal to colnames of counts. "clusters" must be provided (see hierarchy[-1] and notes).</p></dd>


<dt id="arg-hierarchy">hierarchy<a class="anchor" aria-label="anchor" href="#arg-hierarchy"></a></dt>
<dd><p>List of term_set_labels in the Taxonomy ordered from most gross to most fine (e.g., neighborhood, class, subclass, supertype).</p></dd>


<dt id="arg-counts">counts<a class="anchor" aria-label="anchor" href="#arg-counts"></a></dt>
<dd><p>A count matrix in sparse format: dgCMatrix. buildTaxonomy can work with count matrices that have cells are rows or columns, so long as counts has both row names AND column names.</p></dd>


<dt id="arg-highly-variable-genes">highly_variable_genes<a class="anchor" aria-label="anchor" href="#arg-highly-variable-genes"></a></dt>
<dd><p>Set of features defined as highly variable genes OR a number of binary genes to calculate (we recommend ~1000 - ~5000, for &lt;100 to ~5000 cell types). If a feature list is provided, provide either as a named list of vectors, or as a single vector (in which case the name "highly_variable_genes_standard" will be used). "highly_variable_genes_standard" will also be used for calculated variable genes. Optional input, but for proper mapping we strongly recommend including either highly_variable_genes or marker_genes.</p></dd>


<dt id="arg-marker-genes">marker_genes<a class="anchor" aria-label="anchor" href="#arg-marker-genes"></a></dt>
<dd><p>Set of features defined as marker genes. Provide either as a named list of vectors, or as a single vector (in which case the name "marker_genes_mode.name" will be used).</p></dd>


<dt id="arg-ensembl-id">ensembl_id<a class="anchor" aria-label="anchor" href="#arg-ensembl-id"></a></dt>
<dd><p>A vector of ensembl ids corresponding to the gene symbols in counts.</p></dd>


<dt id="arg-gene-meta-data">gene.meta.data<a class="anchor" aria-label="anchor" href="#arg-gene-meta-data"></a></dt>
<dd><p>Either NULL (default) or a data frame of additional gene information to include in the var component of anndata</p></dd>


<dt id="arg-cluster-stats">cluster_stats<a class="anchor" aria-label="anchor" href="#arg-cluster-stats"></a></dt>
<dd><p>A matrix of median gene expression by cluster. Cluster names must exactly match meta.data$cluster.  If provided, will get saved to "varm$cluster_id_median_expr_<a href="https://rdrr.io/r/base/mode.html" class="external-link">mode</a>"</p></dd>


<dt id="arg-embeddings">embeddings<a class="anchor" aria-label="anchor" href="#arg-embeddings"></a></dt>
<dd><p>Dimensionality reduction coordinate data.frame with 2 columns or a string with the column name for marker_genes or variable_genes from which a UMAP should be calculated. If coordinates are provided, rownames must be equal to colnames of counts.  Either provide as a named list or as a single data.frame (in which case the name "default_standard" will be used). embeddings are not required, but inclusion of at least one embedding is strongly recommended.#'</p></dd>


<dt id="arg-number-of-pcs">number.of.pcs<a class="anchor" aria-label="anchor" href="#arg-number-of-pcs"></a></dt>
<dd><p>Number of principle components to use for calculating UMAP coordinates (default=30). This is only used in embeddings corresponds to a variable gene column from which a UMAP should be calculated.</p></dd>


<dt id="arg-dend">dend<a class="anchor" aria-label="anchor" href="#arg-dend"></a></dt>
<dd><p>Existing dendrogram associated with this taxonomy (e.g., one calculated elsewhere). If provided, dend must be a BINARY tree. Can also input a string with the column name for marker_genes or variable_genes from which a dendrogram should be calculated. If NULL or if the function can't figure out what gene set you want, no dendrogram will be calculated. Failure to define a dendrogram will prevent some mapping algorithms from working properly! The default is to build a dendrogram using the first highly_variable_genes or marker_genes set.</p></dd>


<dt id="arg-taxonomydir">taxonomyDir<a class="anchor" aria-label="anchor" href="#arg-taxonomydir"></a></dt>
<dd><p>The location to save Shiny objects, e.g. "/allen/programs/celltypes/workgroups/rnaseqanalysis/shiny/10x_seq/NHP_BG_20220104/"</p></dd>


<dt id="arg-cluster-colors">cluster_colors<a class="anchor" aria-label="anchor" href="#arg-cluster-colors"></a></dt>
<dd><p>An optional named character vector where the values correspond to colors and the names correspond to celltypes in hierarchy[-1].  If this vector is incomplete, a warning is thrown and it is ignored. cluster_colors can also be provided in the metadata (see notes)</p></dd>


<dt id="arg-default-embedding">default_embedding<a class="anchor" aria-label="anchor" href="#arg-default-embedding"></a></dt>
<dd><p>A string indicating which embedding to use for calculations.  Default (NULL) is to take the first one provided in embeddings.</p></dd>


<dt id="arg-uns-variables">uns.variables<a class="anchor" aria-label="anchor" href="#arg-uns-variables"></a></dt>
<dd><p>If provided, a list of additional variables to be included in the uns.  See Notes for schema variables not otherwise accounted for.</p></dd>


<dt id="arg-subsample">subsample<a class="anchor" aria-label="anchor" href="#arg-subsample"></a></dt>
<dd><p>The number of cells to retain per cluster (default = 2000)</p></dd>


<dt id="arg-reorder-dendrogram">reorder.dendrogram<a class="anchor" aria-label="anchor" href="#arg-reorder-dendrogram"></a></dt>
<dd><p>Should dendogram attempt to match a preset order? (Default = FALSE).  If TRUE, the dendrogram attempts to match the celltype factor order as closely as possible (if celltype is a character vector rather than a factor, this will sort clusters alphabetically, which is not ideal).</p></dd>


<dt id="arg-add-dendrogram-markers">add.dendrogram.markers<a class="anchor" aria-label="anchor" href="#arg-add-dendrogram-markers"></a></dt>
<dd><p>If TRUE (default=FALSE), will also add dendrogram markers to prep the taxonomy for tree mapping</p></dd>


<dt id="arg-addmapmycells">addMapMyCells<a class="anchor" aria-label="anchor" href="#arg-addmapmycells"></a></dt>
<dd><p>If TRUE (default), will also prep this taxonomy for hierarchical mapping</p></dd>


<dt id="arg-check-taxonomy">check.taxonomy<a class="anchor" aria-label="anchor" href="#arg-check-taxonomy"></a></dt>
<dd><p>Should the taxonomy be checked to see if it follows the AIT schema (default=TRUE)</p></dd>


<dt id="arg-print-messages">print.messages<a class="anchor" aria-label="anchor" href="#arg-print-messages"></a></dt>
<dd><p>If check.taxonomy occurs, should any messages be written to the screen in addition to the log file (default=TRUE)</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Additional variables to be passed to <code>addDendrogramMarkers</code></p>
<p><em>Additional uns.variables</em>:</p><ul><li><p>dataset_purl: Link to molecular data if not present in X or raw.X.</p></li>
<li><p>batch_condition: Keys defining batches for normalization/integration algorithms. Used for cellxgene.</p></li>
<li><p>reference_genome: Reference genome used to align molecular measurements.</p></li>
</ul></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>AIT anndata object in the specified format (only if return.anndata=TRUE)</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Inkar Kapen.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

